modules/mcc/include/opencv2/mcc/ccm.hpp

```diff
@@ -0,0 +1,493 @@
+        @param gamma the gamma value of gamma correction;
+            NOTICE: only valid when linear is set to "gamma";\n
+            type: double;\n
+            default: 2.2;\n
+            */
+    CV_WRAP void setLinearGamma(double gamma);
+
+    /** @brief set degree
+        @param deg the degree of linearization polynomial;\n
+            NOTICE: only valid when linear is set to "COLORPOLYFIT", "GRAYPOLYFIT",
+                    "COLORLOGPOLYFIT" and "GRAYLOGPOLYFIT";\n
+            type: int;\n
+            default: 3;\n
+            */
+    CV_WRAP void setLinearDegree(int deg);
+        /** @brief set SaturatedThreshold
+        @param lower the lower threshold to determine saturation;\n
+                default: 0;
+        @param upper the upper threshold to determine saturation;
+            NOTICE: it is a tuple of [lower, upper];
+                    The colors in the closed interval [lower, upper] are reserved to participate
+                    in the calculation of the loss function and initialization parameters\n
+                    default: 0;
+            */
+    CV_WRAP void setSaturatedThreshold(double lower, double upper);
+     /** @brief set WeightsList
+        @param weights_list the list of weight of each color;\n
+            type: cv::Mat;\n
+            default: empty array;
+            */
+    CV_WRAP void setWeightsList(Mat weights_list);
+    /** @brief set WeightCoeff
+        @param weights_coeff the exponent number of L* component of the reference color in CIE Lab color space;\n
+            type: double;\n
+            default: 0;\n
+            */
+    CV_WRAP void setWeightCoeff(double weights_coeff);
+    /** @brief set InitialMethod
+        @param initial_method_type the method of calculating CCM initial value;\n
+            Supported list:
+                'LEAST_SQUARE' (least-squre method);
+                'WHITE_BALANCE' (white balance method);\n
+            type: enum INITIAL_METHOD_TYPE;
+            */
+    CV_WRAP void setInitialMethod(INITIAL_METHOD_TYPE initial_method_type);
+    /** @brief set MaxCount
+        @param max_count used in MinProblemSolver-DownhillSolver;\n
+            Terminal criteria to the algorithm;\n
+            type: int;
+            default: 5000;
+            */
+    CV_WRAP void setMaxCount(int max_count);
+     /** @brief set Epsilon
+        @param epsilon used in MinProblemSolver-DownhillSolver;\n
+            Terminal criteria to the algorithm;\n
+            type: double;
+            default: 1e-4;
+            */
+    CV_WRAP void setEpsilon(double epsilon);
+    /** @brief make color correction*/
+    CV_WRAP void run();
+
+    // /** @brief Infer using fitting ccm.
+    //     @param img the input image, type of cv::Mat.
+    //     @param islinear default false.
+    //     @return the output array, type of cv::Mat.
+    // */
+    CV_WRAP Mat infer(const Mat& img, bool islinear = false);
+private:
+    CV_WRAP class Impl;
+    CV_WRAP Ptr<Impl> p;
```

Comment:
> CV_WRAP

No way to wrap private items (no need to do that).

---

Also it is better to keep `Impl` in public section for internal implementation.
```
    class Impl;
private:
    Ptr<Impl> p;
```

---

> Ptr

BTW, on the "master" branch it is allowed to use `std::shared_ptr` for that.

Reply:
Done